### PR TITLE
feat: support write wal logs in columnar format

### PR DESCRIPTION
**Pull request by ShiKaiWi**
## Rationale
After the columnar encoding is supported for wal format, we should adapt such encoding in the wal write procedure.

## Detailed Changes
Adapt the columnar encoding in the wal write procedure. And now, both rowwise and columnar encoding in the wal write procedure are supported, which to use depends on the configuration.

## Test Plan
Existing tests. The compatibility tests have been designed as:
- Write rowwise-encoded wal logs and restart the server to replay it
- Write columnar-encoded wal logs and restart the server to replay it

### CODE REVIEW

**Comment by ShiKaiWi** [comment]
My local test has found that compatibility problem: restart with columnar wal encoding format fails after write some data in rowwise wal encoding format.

File: common_types/src/row/mod.rs [hunk 1]
```diff
@@ -421,6 +394,88 @@ impl<'a> Iterator for IterCol<'a> {
...
+                }
+            }
+        }
+
+        let rows = rows.into_iter().map(Row::from_datums).collect::<Vec<_>>();
+
+        RowGroup {
+            schema: self.schema,
+            rows,
+            min_timestamp: Timestamp::now(),
```
Comment (Rachelint) [inline common_types/src/row/mod.rs:468]: Seems unnecessary to call `now`, and just use a fixed default?

File: common_types/src/datum.rs [hunk 2]
```diff
@@ -1274,6 +1274,13 @@ impl<'a> DatumView<'a> {
         }
     }
 
+    pub fn as_time(&self) -> Option<Timestamp> {
```
Comment (Rachelint) [inline common_types/src/datum.rs:1277]: Just name it `timestamp`?
`Time` may be a specific type as I remembered?

File: analytic_engine/src/payload.rs [hunk 3]
```diff
@@ -163,10 +176,21 @@ pub enum ReadPayload {
 }
 
 impl ReadPayload {
-    fn decode_write_from_pb(buf: &[u8]) -> Result<Self> {
+    fn decode_write_from_pb(schema: Schema, buf: &[u8]) -> Result<Self> {
```
Comment (Rachelint) [inline analytic_engine/src/payload.rs:179]: Seem schema is just used in the `Columnar` path, how about pass the reference and just do the late clone in related path?

File: common_types/src/row/mod.rs [hunk 4]
```diff
@@ -421,6 +394,88 @@ impl<'a> Iterator for IterCol<'a> {
...
+        }
+
+        let old = self.cols.insert(col_id, col);
+        ensure!(old.is_none(), DuplicateColumnId { column_id: col_id });
+
+        Ok(())
+    }
+
+    pub fn build(mut self) -> RowGroup {
+        let num_rows = self.num_rows().unwrap_or(0);
```
Comment (Rachelint) [inline common_types/src/row/mod.rs:431]: Early return in the conercases?

File: components/codec/src/columnar/mod.rs [hunk 5]
```diff
@@ -680,6 +683,72 @@ mod tests {
         check_encode_end_decode(10, datums, DatumKind::Int64);
     }
 
+    #[test]
```
Comment (Rachelint) [inline components/codec/src/columnar/mod.rs:686]: Maybe we should add some cases for the bad case?
Such as zero columns or zero rows?

File: common_types/src/datum.rs (hunk 2 above)
Comment (ShiKaiWi) [reply common_types/src/datum.rs:1277]: `as_timestamp` has been used for the variant `Datum::Timestamp`.

File: analytic_engine/src/payload.rs (hunk 3 above)
Comment (ShiKaiWi) [reply analytic_engine/src/payload.rs:179]: 👍

File: common_types/src/row/mod.rs (hunk 1 above)
Comment (ShiKaiWi) [reply common_types/src/row/mod.rs:468]: Oh. Maybe the correct timestamp should be computed here.

File: common_types/src/datum.rs (hunk 2 above)
Comment (Rachelint) [reply common_types/src/datum.rs:1277]: How about just adding a branch to original `as_timestamp`?

File: common_types/src/row/mod.rs (hunk 1 above)
Comment (Rachelint) [reply common_types/src/row/mod.rs:468]: > Oh. Maybe the correct timestamp should be computed here.

It seems time range is unnecessary actually? Maybe default value is enough?

Comment (ShiKaiWi) [reply common_types/src/row/mod.rs]: Maybe we can't make such an assumption.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Comment by ShiKaiWi** [comment]
50% WAL payload is reduced after this patch applied to the production environment.

**Pull request merged**